### PR TITLE
Support 8-bit floats in the runtime

### PR DESCRIPTION
**Pull request by bjacob**
There are 2 commits in this PR:
1. `math.h` conversions for 8-bit float types. This was previously sent for review as https://github.com/iree-org/iree/pull/17989 but is now folded into this PR.
2. General runtime/HAL support.

### CODE REVIEW

File: runtime/src/iree/hal/string_util.c [hunk 1]
```diff
@@ -16,6 +16,7 @@
 
 #include "iree/base/api.h"
 #include "iree/base/internal/math.h"
+#include "iree/base/status.h"
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: not needed - iree/base/api.h pulls in all iree/base/ headers

File: runtime/src/iree/hal/string_util.c [hunk 2]
```diff
@@ -164,6 +177,36 @@ IREE_API_EXPORT iree_status_t iree_hal_parse_element_type(
 IREE_API_EXPORT iree_status_t iree_hal_format_element_type(
     iree_hal_element_type_t element_type, iree_host_size_t buffer_capacity,
     char* buffer, iree_host_size_t* out_buffer_length) {
+  const char* specialname;
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: always default initialize (don't rely on code 10+ lines away)

File: runtime/src/iree/hal/string_util.c [hunk 3]
```diff
@@ -164,6 +177,36 @@ IREE_API_EXPORT iree_status_t iree_hal_parse_element_type(
...
+      specialname = "f8E4M3";
+      break;
+    case IREE_HAL_ELEMENT_TYPE_FLOAT_8_E5M2_FNUZ:
+      specialname = "f8E5M2FNUZ";
+      break;
+    case IREE_HAL_ELEMENT_TYPE_FLOAT_8_E4M3_FNUZ:
+      specialname = "f8E4M3FNUZ";
+      break;
+    default:
+      specialname = NULL;
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: always break - fallthroughs are no good

File: runtime/src/iree/hal/string_util.c [hunk 4]
```diff
@@ -164,6 +177,36 @@ IREE_API_EXPORT iree_status_t iree_hal_parse_element_type(
...
+  }
+  if (specialname) {
+    int n = strlen(specialname);
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: use proper types: size_t

File: runtime/src/iree/hal/string_util.c [hunk 5]
```diff
@@ -164,6 +177,36 @@ IREE_API_EXPORT iree_status_t iree_hal_parse_element_type(
...
+    if (n + 1 > buffer_capacity) {
+      return iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                              "insufficient buffer capacity");
+    }
+    if (out_buffer_length) {
+      *out_buffer_length = n;
+    }
+    strcpy(buffer, specialname);
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: use memcpy if you've already done the strlen above - or just use snprintf like below to ensure it has matching behavior

File: runtime/src/iree/hal/string_util.c (hunk 2 above)
Comment (benvanik) [inline runtime/src/iree/hal/string_util.c]: `special_name`

File: runtime/src/iree/hal/string_util_test.cc [hunk 6]
```diff
@@ -14,6 +14,7 @@
 #include "iree/base/api.h"
 #include "iree/base/internal/span.h"
 #include "iree/hal/api.h"
+#include "iree/hal/buffer_view.h"
```
Comment (benvanik) [inline runtime/src/iree/hal/string_util_test.cc]: don't include - iree/hal/api.h has it

(if you have your editor doing this, turn it off)

**Review by benvanik — CHANGES_REQUESTED** [review]

File: runtime/src/iree/hal/buffer_view.h [hunk 7]
```diff
@@ -48,6 +48,13 @@ enum iree_hal_numerical_type_bits_t {
   IREE_HAL_NUMERICAL_TYPE_FLOAT_BRAIN = IREE_HAL_NUMERICAL_TYPE_FLOAT | 0x02u,
   // Paired (real, imag) complex number in floating-point format.
   IREE_HAL_NUMERICAL_TYPE_FLOAT_COMPLEX = IREE_HAL_NUMERICAL_TYPE_FLOAT | 0x03u,
+  // Ad-hoc entries for the zoo of low-bit-depth float types. There currently
+  // are 4 8-bit float types that we need to support. For now just create these
+  // 4 nameless "slots" to fit those types.
+  IREE_HAL_NUMERICAL_TYPE_FLOATISH_0 = IREE_HAL_NUMERICAL_TYPE_FLOAT | 0x04u,
```
Comment (benvanik) [inline runtime/src/iree/hal/buffer_view.h]: "FLOATISH" is cute but we want something more precise here. If these are pinned to specific encodings then just name them the specific encoding (no need for a mapping)

File: runtime/src/iree/hal/string_util.c (hunk 1 above)
Comment (bjacob) [reply runtime/src/iree/hal/string_util.c]: ah yeah, the auto header includes are an annoying default feature in vscode/clangd. I disable it when I think about it, but here I'm on a different machine for the GPU.

File: runtime/src/iree/hal/string_util.c (hunk 2 above)
Comment (bjacob) [reply runtime/src/iree/hal/string_util.c]: i know! i was trying to imitate local style (see `const char* prefix;` below in this function).

**Review by benvanik — APPROVED** [review]

**Pull request merged**